pcicrawler/ocp_diag.py

```diff
@@ -0,0 +1,259 @@
+
+    def run(self):
+        return super().run(self._device.express_aer)
+
+
+class TestSet(object):
+    '''Object to hold a list of DUTs to be tested and 
+    the values they need to be compared against.'''
+    def __init__(self, 
+                 duts: ty.List[PCIDevice], 
+                 validate:ty.Dict[str, ty.Union[int, str, ty.Dict[str, str]]]):
+        self.duts = duts
+        self.conditions = validate
+
+    def __str__(self):
+        return str(self.duts)
+
+
+# List of available tests to iterate through
+AVAILABLE_TESTS = {
+    'location': CheckLocation,
+    'vendor_id': CheckVendorID,
+    'device_id': CheckDeviceID,
+    'class_id': CheckClassID,
+    'physical_slot': CheckPhysicalSlot,
+    'current_link_speed': CheckCurrentLinkSpeed,
+    'current_link_width': CheckCurrentLinkWidth,
+    'capable_link_speed': CheckCapableLinkSpeed,
+    'capable_link_width': CheckCapableLinkWidth,
+    'addresses': CheckAddress,
+    'check_aer': CheckAER,
+}
+
+
+class OCPOutputObj:
+    def __init__(self,
+                 devs: ty.Set[PCIDevice],
+                 json_path: str,
+                 ocp_run: tv.TestRun):
+        self._json_path = json_path
+        self._devs = devs
+        self._ocp_run = ocp_run
+        self._result = True
+
+    def run(self):
+        duts = []
+        # Process input json
+        with open(self._json_path, 'r') as fh:
+            input_json = json.load(fh)
+            try:
+                duts = input_json['duts']
+            except KeyError as _:
+                raise json.JSONDecodeError("ERROR: Cannot parse file {}, missing DUTs info.".format(self._json_path))
+        filtered_duts = []
+        for dut in duts:
+            identifier = dut['identifiers']
+            potential_duts = list(self._devs)
+            # Go look for this device and get the data
+            missing_devices = []
+            if 'address' in identifier:
+                if identifier['address'] not in potential_duts:
+                    missing_devices.append("No device was found with address {}.".format(identifier['address']))
+                potential_duts = [dut for dut in potential_duts if str(dut) == identifier['address']]
+            if 'vendor_id' in identifier:
+                potential_duts = [dut for dut in potential_duts if dut.vendor_id == identifier['vendor_id']]
+                if len(potential_duts) == 0:
+                    missing_devices.append("No device was found with vendor id {}.".format(identifier['vendor_id']))
+            if 'device_id' in identifier:
+                potential_duts = [dut for dut in potential_duts if dut.device_id == identifier['device_id']]
+                if len(potential_duts) == 0:
+                    missing_devices.append("No device was found with device id {}.".format(identifier['device_id']))
+            if len(potential_duts) != 0:
+                filtered_duts.append(TestSet(potential_duts, dut['validate']))
```

Comment:
it's unclear to me how this input json was established. why does it both contain data about devices and filtering criteria? why isnt the input json _already_ filtered?

is this documented somewhere? readme? json schema?

Reply:
i found the answer for the filtering part in a _resolved_ comment from my first review (which was difficult to see because it was collapsed due to being resolved; to remove ambiguity, i think it's the reviewer's role to mark comments as resolved after checking that the initial reported issue was fixed)

the other questions remain

Reply:
I'm making an assumption that people don't necessarily follow the rules to a T lol

Reply:
that introduces confusion. in any case, what about my initial question? i still dont see an answer
> why does it both contain data about devices and filtering criteria? why isnt the input json already filtered?
> is this documented somewhere? readme? json schema?